### ISSUE TITLE
Double execution on started server and on script that adds a task

### ISSUE THREAD

**Issue by milanbgd011** [resonatehq/resonate-sdk-ts#146]
## Expected Behavior

The function that adds task not to return anything, then server to return the "console.log" outputs.

## Actual Behavior

The "server-addtask.js" outputs the console logs with timestamps (because I did not use "await" I did not expect anything to be console logged), but that are COMPLETELY DIFFERENT from what the server is outputing as timestamps, pointing to the conclusion that this code is executed actually 2 times? It would output same timestamp if it is same execution.

Looking at the docs and "ask echo", if I use:
- `resonateInstance.run("f1", id, 'one', 'two')` - it should add task and not output console.logs
- `await resonateInstance.run("f1", id, 'one', 'two')` - it should wait for task to be executed and output console.logs

<img width="1512" alt="Screenshot 2024-10-02 at 22 08 27" src="https://github.com/user-attachments/assets/a40a6bd3-ece8-4ce8-a88e-b2f4f14a5277">

1 > on screenshot (number 1) is the output of the server
2 > on screenshot (number 2) is the output of the function that adds a tas

## To Reproduce

1. start server with `resonate serve` (installed thru brew)

2. create server.js

```js
import { Resonate, Context } from "@resonatehq/sdk";
export const resonateInstance = new Resonate({
    url: "http://localhost:8001",
});

async function f4(ctx, user, song) {
  console.log('')
  console.log('4 PRE', new Date());
  await new Promise(r => setTimeout(r, 3_000));
  console.log('4 POST', new Date());
  return {}
}


async function f3(ctx, user, song) {
  console.log('')
  console.log('3 PRE', new Date());
  await new Promise(r => setTimeout(r, 3_000));
  console.log('3 POST', new Date());
  return {}
}


async function f2(ctx, user, song) {
  console.log('')
  console.log('2 PRE', new Date());
  await new Promise(r => setTimeout(r, 3_000));
  console.log('2 POST', new Date());
  return {}
}

async function f1(ctx, user, song) {
  console.log('')
  const charged = await ctx.run(f2, user, song);
  const granted = await ctx.run(f3, user, song);
  const wtf = await ctx.run(f4, user, song);
  return { charged, granted, wtf };
}

// 2) Register purchase as a durable async function
resonateInstance.register("f1", f1, resonateInstance.options({ timeout: 60_000 }));
resonateInstance.register("f2", f2, resonateInstance.options({ timeout: 60_000 }));
resonateInstance.register("f3", f3, resonateInstance.options({ timeout: 60_000 }));
resonateInstance.register("f4", f4, resonateInstance.options({ timeout: 60_000 }));


await resonateInstance.start()
```

3.  create server-addtask.js

```js
import { resonateInstance } from './server.js'
const id = +new Date() // intentionally to get new id when starting new task for development only
resonateInstance.run("f1", id, 'one', 'two')

```

4. start server

`deno  run --allow-all server.js`

5. add task

`deno  run --allow-all server-addtask.js`

6. observe the timestamps

## Specifications

- Version: 0.6
- Platform: macos

**Comment by avillega** [comment]
Hey @milanbgd011, I am not sure I am following, but let me try to explain what might be happening.
You mentioned
> (because I did not use "await" I did not expect anything to be console logged)

In Javascript, execution of an async function starts when you invoke it, not when you await it. Awaiting allows you to wait for the return value of such function, but the function might be executing in the background since the moment you call it. I understand the confusion because in some languages, for example rust, execution do not start until you do await. 

We also have introduced structured concurrency into our runtime which means that before exiting a `Context` we make sure all executions inside that `Context` are finished. This is not what is happening here, but something similar. Deno and other runtimes also wait for all the executions to finish before exiting the process.

Lastly, since you are giving a new id to the second execution, we treat it as a new execution and therefore run the function again. If you want a "cached" result you can use the same id you used in the first execution and we will return the resolved value of the promise instead of re-execution the function a second time.

Please let me know if this addresses your problems or not.

**Comment by milanbgd011** [comment]
@avillega Hi, to introduce myself, I was interested in Temporal initially and was trying to make something that was clear and usable, but it was really bad abstraction. Here is me talking to Lorensr (main TS guy in temporal): https://github.com/temporalio/samples-typescript/issues/97

Ok, here is better example, I am updating the SQLlite database counter +1 on each run.
You can clearly see that this code is being executed 2 times because of the counter increase on both server and client JS.

<img width="670" alt="Screenshot 2024-10-03 at 20 29 03" src="https://github.com/user-attachments/assets/165d7a3c-70b3-411d-92d1-0f8f69bc76a7">

I am using "node" now and you can reproduce on your machine easy, just create 2 files and run them.

### server.js

```js
import { Resonate, Context } from "@resonatehq/sdk";

import sqlite3 from 'sqlite3';
const db = new sqlite3.Database('catcounter.db');

function initializeDatabase() {
  return new Promise((resolve, reject) => {
    db.run(`CREATE TABLE IF NOT EXISTS catcounter (
      id INTEGER PRIMARY KEY,
      countcats INTEGER DEFAULT 0
    )`, (err) => {
      if (err) reject(err);
      db.run(`INSERT OR IGNORE INTO catcounter (id, countcats) VALUES (1, 0)`, (err) => {
        if (err) reject(err);
        resolve();
      });
    });
  });
}

function updateCatCount() {
  return new Promise((resolve, reject) => {
    db.run(`UPDATE catcounter SET countcats = countcats + 1 WHERE id = 1`, (err) => {
      if (err) reject(err);
      resolve();
    });
  });
}

function readCatCount() {
  return new Promise((resolve, reject) => {
    db.get(`SELECT countcats FROM catcounter WHERE id = 1`, (err, row) => {
      if (err) reject(err);
      resolve(row ? row.countcats : 0);
    });
  });
}

await initializeDatabase();


async function f4(ctx, user, song) {
  console.log('');
  console.log('4 PRE', new Date());
  await new Promise(r => setTimeout(r, 3000));
  console.log('4 POST', new Date());
  await updateCatCount();
  console.log(`Current cat count: ${await readCatCount()}`);
  return {};
}

async function f3(ctx, user, song) {
  console.log('');
  console.log('3 PRE', new Date());
  await new Promise(r => setTimeout(r, 3000));
  console.log('3 POST', new Date());
  await updateCatCount();
  console.log(`Current cat count: ${await readCatCount()}`);
  return {};
}

async function f2(ctx, user, song) {
  console.log('');
  console.log('2 PRE', new Date());
  await new Promise(r => setTimeout(r, 3000));
  console.log('2 POST', new Date());
  await updateCatCount();
  console.log(`Current cat count: ${await readCatCount()}`);
  return {};
}

async function f1(ctx, user, song) {
  console.log('');
  await updateCatCount();
  console.log(`Current cat count: ${await readCatCount()}`);
  const charged = await ctx.run(f2, user, song);
  const granted = await ctx.run(f3, user, song);
  const wtf = await ctx.run(f4, user, song);
  return { charged, granted, wtf };
}


export const resonateInstance = new Resonate({
  url: "http://localhost:8001",
});

// Register functions
resonateInstance.register("e1", f1, resonateInstance.options({ timeout: 60000 }));
resonateInstance.register("e2", f2, resonateInstance.options({ timeout: 60000 }));
resonateInstance.register("e3", f3, resonateInstance.options({ timeout: 60000 }));
resonateInstance.register("e4", f4, resonateInstance.options({ timeout: 60000 }));

// Start the instance
await resonateInstance.start();

```

### server-addtask.js

```js
import { resonateInstance } from './server.js'

const id = 'dfasdfd'

console.log('start')
resonateInstance.run("e1", id, 'one', 'two')
console.log('done')
// process.exit(0)
```

**Comment by milanbgd011** [comment]
@avillega For example I used this in Bull npm package and it works intuitively

### server.js (pseudo code)

```js
const Queue = require('bull');
const myQueue = new Queue('my-queue', 'redis://127.0.0.1:6379');
myQueue.process(async (job) => {
  console.log('Processing job:', job.data);
  // Simulate some work
  await new Promise(resolve => setTimeout(resolve, 5000));
  console.log('Job completed:', job.data);
  return { result: 'Job processed successfully' };
});
```

### server-addtask.js (pseudo code)

```js
const Queue = require('bull');
const myQueue = new Queue('my-queue', 'redis://127.0.0.1:6379');
await myQueue.add(data);
```

This works perfectly, no problems. Same setup in Resonate is behaving weird as demonstrated.
Maybe I am misunderstanding something, I read the docs and all. Tried all but failed.

**Comment by avillega** [comment]
Thanks for the examples! The Queue examples you show is quite intuitive when thinking about jobs. In resonate, it is totally possible to process jobs but I will encourage you to think about them as just simple async functions. I rewrote your first example using bun as a run time and changing things a bit to show the behaviour more explicitly:
## server.ts
```ts
import { Database } from "bun:sqlite";
import { setTimeout } from "timers/promises";
import { Resonate, Context } from "@resonatehq/sdk";

const db = new Database(":memory:");

async function initializeDatabase() {
  console.log("initializing db.");

  db.run(`CREATE TABLE IF NOT EXISTS catcounter (
    id INTEGER PRIMARY KEY,
    countcats INTEGER DEFAULT 0
  )`)

  db.run(`INSERT OR IGNORE INTO catcounter (id, countcats) VALUES (1, 0)`)

  console.log("db initialized");
}

async function updateCatCount(): Promise<void> {
    db.run(`UPDATE catcounter SET countcats = countcats + 1 WHERE id = 1`);
    await setTimeout(100);
}

async function readCatCount(): Promise<number> {
  let query = db.query('SELECT countcats FROM catcounter WHERE id = 1');
  await setTimeout(100);
  const {countcats} = query.all()[0] as {countcats: number}
  return countcats
}

await initializeDatabase()

export const resonate = new Resonate()

async function f4(ctx:Context) {
  console.log('');
  console.log('4 PRE', new Date());
  await setTimeout(3000);
  console.log('4 POST', new Date());
  await updateCatCount();
  return await readCatCount();
}

async function f3(ctx: Context) {
  console.log('');
  console.log('3 PRE', new Date());
  await setTimeout(3000)
  console.log('3 POST', new Date());
  await updateCatCount();
  return await readCatCount();
}

async function f2(ctx: Context) {
  console.log('');
  console.log('2 PRE', new Date());
  await setTimeout(3000)
  console.log('2 POST', new Date());
  await updateCatCount();
  return await readCatCount();
}

async function f1(ctx: Context): Promise<any> {
  console.log('Starting f1');
  await updateCatCount();
  console.log(`Current cat count: ${await readCatCount()}`);
  const first = await ctx.run(f2);
  const second = await ctx.run(f3);
  const third = await ctx.run(f4);
  return { first,second, third } as any;
}


// Register functions
resonate.register("e1", f1, resonate.options({ timeout: 60000 }));
await resonate.start()
``` 

## index.ts
```ts
import { resonate } from './server.ts'

const id = 'dfasdfd'

console.log('starting first run of e1 function.')
const a: any  = await resonate.run("e1", id)
console.log("After first resonate.run", a)

console.log("Runing again")
const b: any  = await resonate.run("e1", id)
console.log("After second resonate run", b)
console.log("Note that the PRE/POST lines are not being log because we are caching the result")

console.log("Third time with a totally different id")
const c: any  = await resonate.run("e1", "lalal")
console.log("After third resonate run", c)
console.log("Note that the PRE/POST lines ARE log because we are starting a new execution")
console.log("also the catcounts have incresed because we ran the queies again")

console.log('done')

resonate.stop()
```

A few things to note. 
- I am not passing an `url` to the resonate instance, I am running in local mode, which simulates the promise store but in memory.
- I am using setTimout promise instead of wrapping it in a promise, also so simulate some work.
- the bun sqlite api is synchronous, I am adding some setTimeouts to made them async
- I am only registering the function I am calling through `resonate.run` the other functions do not need to be registered. 

## The output of the execution my sample is:
```
initializing db.
db initialized
starting first run of e1 function.
Starting f1
Current cat count: 1

2 PRE 2024-10-03T21:53:56.065Z
2 POST 2024-10-03T21:53:59.068Z

3 PRE 2024-10-03T21:53:59.280Z
3 POST 2024-10-03T21:54:02.283Z

4 PRE 2024-10-03T21:54:02.492Z
4 POST 2024-10-03T21:54:05.495Z
After first resonate.run {
  first: 2,
  second: 3,
  third: 4,
}
Runing again
After second resonate run {
  first: 2,
  second: 3,
  third: 4,
}
Note that the PRE/POST lines are not being log because we are caching the result
Third time with a totally different id
Starting f1
Current cat count: 5

2 PRE 2024-10-03T21:54:05.908Z
2 POST 2024-10-03T21:54:08.910Z

3 PRE 2024-10-03T21:54:09.117Z
3 POST 2024-10-03T21:54:12.118Z

4 PRE 2024-10-03T21:54:12.323Z
4 POST 2024-10-03T21:54:15.324Z
After third resonate run {
  first: 6,
  second: 7,
  third: 8,
}
Note that the PRE/POST lines ARE log because we are starting a new execution
also the catcounts have incresed because we ran the queies again
done
```

**Comment by milanbgd011** [comment]
@avillega Thank you for the example. Works as you described. I will now dig a bit deeper into it and give feedback. I would take couple of days for me to test all different scenarios I have in my mind.

You can remove "bug" label, it is not a bug definitely.

Now I am even more excited about Resonate :)

**Comment by avillega** [comment]
Awesome, please feel free to create more issues with any feedback, suggestions or feature request, I will close this one for now.
